Marlin/src/feature/pause.cpp

```diff
@@ -252,94 +256,182 @@ void wait_for_filament_reload(const int8_t max_beep_count/*=0*/) {
+	
+  // Stop blowing after delay
+  #ifdef PAUSE_DWELL
+    delay(PAUSE_DWELL);
+  #endif
+  #if (PAUSE_FANSPEED >0 && PAUSE_FAN < FAN_COUNT)
+    fanSpeeds[PAUSE_FAN]=0 ;
+  #endif  
+  
+  #ifdef PAUSE_RET
+  // Intelligent Retractation before return to print
+  // If resume position negative then travel back Retraction = resume_position
+  // Rare but possible : can go on pause with retracted/swapped status or negative position
+  #ifdef PAUSE_RET
+    #if ENABLED(FWRETRACT)
+	  //If retracted/swapped goto negative position
+	  if (fwretract.retracted_swap[active_extruder]) {
+	    const float tmp_length=fwretract.swap_retract_length;
+	    const float tmp_feed=fwretract.retract_feedrate_mm_s;
+	    do_pause_e_move(tmp_length, tmp_feed); 
+	  }
+	  // If retracted or not : normal fwretract used
+	  else {
+	    const float tmp_length=fwretract.retract_length;  
+            const float tmp_feed=fwretract.retract_feedrate_mm_s;   
+            do_pause_e_move(tmp_length, tmp_feed); 
+	  }
+    #else
+	  //If negative and more than PAUSE_RET goto resume_position 
+	  if (resume_position[E_AXIS]<(-PAUSE_RET)) do_pause_e_move(resume_position[E_AXIS], PAUSE_RET_F);
+          else do_pause_e_move(-PAUSE_RET, PAUSE_RET_F);
+    #endif
+  #endif	
+ 
   // Move XY to starting position, then Z
-  do_blocking_move_to_xy(resume_position[X_AXIS], resume_position[Y_AXIS], PAUSE_PARK_XY_FEEDRATE);
-  do_blocking_move_to_z(resume_position[Z_AXIS], PAUSE_PARK_Z_FEEDRATE);
-
-  #if ENABLED(FILAMENT_RUNOUT_SENSOR)
+  do_blocking_move_to_xy(resume_position[X_AXIS], resume_position[Y_AXIS], PAUSE_XY_F);
+
+   //Set Z_AXIS to saved position  
+  do_blocking_move_to_z(resume_position[Z_AXIS], PAUSE_Z_F);
+  
+  // Intelligent Retractation Recovery
+  // Keep applied retractation as resume position if retracted before goto pause
+  #ifdef PAUSE_RET
+    #if ENABLED(FWRETRACT)
+	  //if not swapped before , then recover to zero
+	  if (!fwretract.retracted_swap[active_extruder] ){
+	    const float tmp_length=fwretract.swap_retract_recover_length;
+	    const float tmp_feed=  fwretract.swap_retract_recover_feedrate_mm_s;
+	    do_pause_e_move(tmp_length, tmp_feed); 
+	  }
+	  //if not retracted before , then recover to zero
+	  else if (!fwretract.retracted[active_extruder]) {
+	         const float tmp_length=fwretract.retract_recover_length;
+	         const float tmp_feed=  fwretract.retract_recover_feedrate_mm_s;
+                 do_pause_e_move(tmp_length, tmp_feed);
+	       }			
+	#else    
+	  // If resume_position negative goto resume_position
+      if (resume_position[E_AXIS]<0) do_pause_e_move(resume_position[E_AXIS], PAUSE_RET_F);
+	  // if positive soft recovery to zero for sure print restart
+      else do_pause_e_move(PAUSE_RET,15); 
+    #endif
+  #endif
+  
+  //Now all extrusion positions are resumed and ready to be confirmed
+  //Set extruder to saved position
+  planner.set_e_position_mm((current_position[E_AXIS] = resume_position[E_AXIS]));
+  
```

Comment:
Not it's time to recover , if  retracted/swapped or negative before go on pause , i move E to the exact negative positin , to be ready on the return to print
If not , i reconver with a soft standard value , because , if no fw , it's impossible to have an idea of the retractation feedrate , and it's VERY IMPORTANT , too quick = hole in extrusion restart , too long= strong line 
15 mms second is perfect and sure